processes/fulfill_url_manifest.py

```diff
@@ -0,0 +1,166 @@
+                    ContentType = 'application/json'
+                )
+            except ClientError as e:
+                logging.error(e)
+
+    def link_fulfill(self, metadata_json, counter):
+        for link in metadata_json['links']:
+            fulfill_link, counter = self.fulfill_replace(link, counter)
+            link['href'] = fulfill_link
+
+        return (metadata_json, counter)
+            
+    def reading_order_fulfill(self, metadata_json, counter):
+        for read_order in metadata_json['readingOrder']:
+            fulfill_link, counter = self.fulfill_replace(read_order, counter)
+            read_order['href'] = fulfill_link
+
+        return (metadata_json, counter)
+
+    def resource_fulfill(self, metadata_json, counter):
+        for resource in metadata_json['resources']:
+            fulfill_link, counter = self.fulfill_replace(resource, counter)
+            resource['href'] = fulfill_link
+
+        return (metadata_json, counter)
+
+    def toc_fulfill(self, metadata_json, counter): 
+
+        '''
+        The toc dictionary has no "type" key like the previous dictionaries 
+        therefore the 'href' key is evaluated instead
+        '''
+
+        for toc in metadata_json['toc']:
+            if 'pdf' in toc['href'] \
+                or 'epub' in toc['href']:
+                    for link in self.session.query(Link) \
+                        .filter(Link.url == toc['href'].replace('https://', '')):
+                            counter += 1
+                            toc['href'] = f'https://{self.host}/fulfill/{link.id}'
+
+        return (metadata_json, counter)
+
+    def fulfill_replace(self, metadata, counter):
+        if metadata['type'] == 'application/pdf' or metadata['type'] == 'application/epub+zip' \
+            or metadata['type'] == 'application/epub+xml':
+                for link in self.session.query(Link) \
+                    .filter(Link.url == metadata['href'].replace('https://', '')):
+                            counter += 1            
+                            metadata['href'] = f'https://{self.host}/fulfill/{link.id}'
+
+        return (metadata['href'], counter)
+    
+    def fulfill_flag_update(self, metadata):
+        if metadata['type'] == 'application/webpub+json':
+            for link in self.session.query(Link) \
+                .filter(Link.url == metadata['href'].replace('https://', '')):   
+                        if 'fulfill_limited_access' in link.flags.keys():
+                            if link.flags['fulfill_limited_access'] == False:
+                                newLinkFlag = dict(link.flags)
+                                newLinkFlag['fulfill_limited_access'] = True
+                                link.flags = newLinkFlag
+                                self.commitChanges()
+                        
+    def check_rights_status(self, metadata_json, rights_status):
+        for metadata in metadata_json['links']:
```

Comment:
are we iterating over the metadata or or the links

Reply:
We're iterating over the links. I just changed it because I had two 'link' variables in this method with one representing the manifest link and another for a database link object. I'll change this `metadata` variable to `link` and the database link object to `psql_link`